### PR TITLE
Fix apply default values on update behavior

### PR DESCRIPTION
**Pull request by signedav**
## 1. apply-on-update-default-values where always emitted in attribute table #54446
**Problem:**
Even on showing them - even when the edit mode was off. This leaded to low performance and displayed values that are not in the database like this.

**Solution:**
Now, we do not set the default values in the widgets on apply-default-values-on-update on `resetValues` anymore. Instead we only update those values when a depending field changed. (Virtual field values are still emitted)

## 2. apply-on-update-default-values where not emitted when they where not depending on other fields #53722
**Problem:**
This was made on purpose not to impact the performance to much. Still without the bug fixed above it used to be super confusing that values where updated, but not visible in the widget (just in the background). This was mostly noticed on values depending on "time", why it made sense to consider expressions containing `now()`.

**Solution:**
As a compromise solution - not to impact the performance too much what would happen when we update every widget containing a function, we collect all the widgets containing the `now()`-function in the default value expression as well as other volatile functions like `rand`, `randf` and `uuid`, and if so, we do update the apply-on-update-default-values on every field value change (even when it's not a dependency). Same on virtual field values.

**Additional info:**
It needs to be discussed, what kind of functions should be considered. 
Btw. all functions requiring `ALL_ATTRIBUTES` are considered anyway - what can be good or bad. See still existing problem [ in this comment](https://github.com/qgis/QGIS/pull/55633#issuecomment-1900696811)

## 3. Concern geometry in expressions

**Problem:**
After those fixes where made another issue appeared. `$area`, `$geometry` etc. needed the geometry in the expressions, those where not concerned in the cached features. 

**Solution:**
Now we check the attribute form if it needs geometry and if so, we cache it as well.

### Add Feature 
- Create and form opens:
  - normal-default-values are set
  - apply-on-update-default-values are set
- Changing a value in the open form: 
  - normal-defalt-values are *not* reset
  - apply-on-update-default-values are reset \*
- Pressing "Ok" to commit feature or "Cancel" to discard feature

### Identify and open Form
- Form opens:
  - normal-default-values are *not* set
  - apply-on-update-default-values are *not* set
- Changing a value in the open form (only possible in edit mode): 
  - normal-defalt-values are *not* reset
  - apply-on-update-default-values are reset \*
- Pressing "Ok" what commits feature if changes has been made or "Cancel" to discard changes (manual changes and default value changes)

### Attribute table form view
- Selecting feature and form is visible:
  - normal-default-values are *not* set
  - apply-on-update-default-values are *not* set
- Changing a value in the open form (only possible in edit mode): 
  - normal-defalt-values are *not* reset
  - apply-on-update-default-values are reset \*
- "Changing feature selection" what commits feature if changes has been made

### Attribute table table view (unusual use case)
- Changing a value in the open form (only possible in edit mode): 
  - normal-defalt-values are *not* reset
  - apply-on-update-default-values are reset \*
- "Changing feature selection" what commits feature if changes has been made

### Changing geometry
  - apply-on-update-default-values are reset
  
\* concerning the live-update it means that we "see" the changes that are made in the background in the current field-widgets: But there we see *only* the changes, when a field value changed manually that has impacts on this value or when a relevant function (currently `now()` , `rand`, `randf` and `uuid`) is involved. Other values like when e.g. a value changes in the background that has impact on another apply-on-update-default-value, are not updated real-time for the sake of performance. 

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

 This should be backported in my opinion.
 
------------------------

  - [ ] New unit tests have been added for relevant changes

### CODE REVIEW

**Comment by signedav** [comment]
@andreasneumann Not sure if you still have crazy projects to test this. But it would be much appreciated. It's for sure fixing bad behavior but since it's "removing" something it would be good to rule out unexpected behavior on side cases.

**Review by domi4484 — COMMENTED** [review]
Code looks good to me. I will test the behavior in the next few days.

**Comment by semriemens** [comment]
Hi, since you're working on this code, I have a question related to this.
There is some unintuitive behaviour when "Apply default values on update" is set. When identifying an existing feature, default values override values in the feature form, not just when dependent fields change, but **again when OK is pressed**.  So if the user changed any of these attributes their input values are overwritten and they won't find out unless they reopen the feature form.

The same is **not** true when adding a new feature: default values are filled in and updated by fields dependencies but if the user then edits that attribute in the form, the user-specified value is saved. In my opinion that is also the correct behaviour.

With your PR, is this behaviour now fixed? Or is it easy to do so? If not I can open a new bug report.

**Comment by signedav** [comment]
Thanks @semriemens I've noticed that as well. It's not part of this PR. Neither a fix nor it's making it worse.

The problem is that we set the widget values (what sets the attribute values indirectly) and we set the attribute values additionally directly when a feature is updated. Later is needed e.g. on geometry update (not sure if only...)

Anyway. I suggest you to open an issue as soon as this PR is merged (not to mix up things). I would then be happy to look into it as soon as I have the ressources.

**Comment by domi4484** [comment]
@signedav I tried out this branch, and I think something is wrong if the default value expression depends on another field.

I prepared a layer with following config:

| Field name | Default value exp | Apply on update |
| -------- | ------- |  ------- |
| NormalValue                                            |   |  |
| NormalDefaultValue                     | ` "NormalValue" \|\| ' - ' \|\|  now() ` |
| AlwaysUpdateDefault                   | ` "NormalValue" \|\| ' - ' \|\|  now() ` | ✅
| DefaultValueNow                           | `now()` |
| AlwaysUpdateDefaultValueNow | `now()` | ✅

- `DefaultValueNow` and `AlwaysUpdateDefaultValueNow` are updated as expected. 
- When creating a new feature `NormalDefaultValue` is not updated. Even if the depending value `NormalValue` is changed.
- When creating a new feature `AlwaysUpdateDefault` is not updated when the form is opened (stays NULL and I think should be ` - 2023-12-29...`). 

Here my test [project](https://github.com/qgis/QGIS/files/13794755/DefaultValuesTest.qgz.zip)

**Comment by signedav** [comment]
Thanks @domi4484 I'll have a look next week.

**Comment by signedav** [comment]
@domi4484 Not sure if I understand it correct but in my opinion the behavior is right.

It just looks wrong, since your expression `"NormalValue" || ' - ' ||  now()` is NULL, when `NormalValue` is NULL. 

Test it with:
image

(this behavior is okay or not, but has nothing to do with this implementation)

It means on creating a new feature `NormalValue` is NULL and with it `NormalDefaultValue` and `AlwaysUpdateDefault` as well NULL. On changing `NormalValue` the `AlwaysUpdateDefault` gets a value but the `NormalDefaultValue` not.

This kind of expression (concating NULL values) can be implemented like this:
```
COALESCE( "NormalValue", '') || ' - ' ||  now() 
```

I guess here is the behavior like expected:
value

Could you please have another look at it?

**Comment by domi4484** [comment]
> It just appears wrong, since your expression `"NormalValue" || ' - ' || now()` is always NULL, when `NormalValue` is NULL.

Oh sorry you are right. Then this is as expected.

The other thing, I erroneously thought (again) that if `Apply on update`  is not checked, default value should be updated as long as the creating feature form is open. And not just at the moment of opening it, which I think is preferable.

So on my side everything green

**Comment by signedav** [comment]
I removed the functionality that the value is updated on any change as long it contains a function in the expression, for the sake of the performance. Still, since `now()` is used pretty often, it concerns `now()`. Maybe there can be other functions in future.

**Comment by signedav** [comment]
Still we should think about other improvements - but not in this PR. Let this PR only fix the most important things.

Anyway I note it here and if this PR is merged I will create an issue for it:

When is a value evaluated on apply on update for a widget:
- When changing a field that current widgets expression depends on
- When changing any field and the current widget expression contains the function `now()`
- When changing any field and the current widget expression contains a function using `ALL_ATTRIBUTES`

Especially the third thing does only make partly sense, but I don't see a better solution for it. Maybe it would make sense to change all the fields on every change and have a setting to turn off live update per widget to improve performance.

File: src/gui/qgsattributeform.cpp
Comment (nirvn) [inline src/gui/qgsattributeform.cpp:3158]: This feels a bit ad-hoc, and won't scale up if other functions have similar requirements. I'm thinking for e.g. uuid(), rand(), randf(). 

Wondering if that could all be regrouped into maybe a static QgsExpressionUtils function where such a list could be maintained.

**Comment by nirvn** [comment]
@signedav , code looks good, and thanks for having opted for to new tests, makes the one tiny behavior fix on the preexisting test much easier to understand.

File: src/gui/qgsattributeform.cpp
Comment (m-kuhn) [reply src/gui/qgsattributeform.cpp:3158]: Ideally we would flag these functions, similar to [`isStatic()`](https://github.com/qgis/QGIS/blob/2e7c0e6b32e0d39f72463ae4651b42109e0a2de3/src/core/expression/qgsexpressionfunction.h#L244), (e.g. `isVolatile()`) instead of maintaining a list. If someone registers a python based function they should also be able to make it behave this way.
Ideally `referencedFunctions()` (or a sibling method) would already return a list of `QgsExpressionFunction` but since this is not the case a helper function in `QgsExpressionUtils` that proxies to this can be an easier option.

Comment (signedav) [reply src/gui/qgsattributeform.cpp:3158]: Consider that this is about the "live" update of the widget value in the form and not if it updates at all. And neither it has something go do with everything besides the default-value-on-update (like constraints etc). So I'm not sure if a flag is the right thing and neither if I would have it in the ExpressionUtils. In my opinion the attribute form is the better place...

Comment (signedav) [reply src/gui/qgsattributeform.cpp:3158]: But I agree to have a static function for it...

Comment (m-kuhn) [reply src/gui/qgsattributeform.cpp:3158]: In this case, we could keep it accessible via python API of the feature form (in this case it would be a non static function of the form to get and set the list of volatile function names)?

Comment (signedav) [reply src/gui/qgsattributeform.cpp:3158]: And where it should be set? On every attributeform call?

Comment (m-kuhn) [reply src/gui/qgsattributeform.cpp:3158]: I would initialize in the constructor (but leave it to python code to override)

Comment (signedav) [reply src/gui/qgsattributeform.cpp:3158]: Hmm... That sounds reasonable, but I'm a little unsure. As I mentioned in this PR, this fixes some bad behavior here, but it may not be the final solution. Maybe every feature should be considered and it should be controlled by a user setting whether this specific widget should be live-updated or not. So I'm a little afraid that with your suggestion we are introducing something that we will soon have to mark as deprecated....

File: src/gui/qgsattributeform.h
Comment (nirvn) [inline src/gui/qgsattributeform.h:359]: \since ...

**Review by nirvn — APPROVED** [review]
Good to merge when the documentation comment is addressed.

**Comment by signedav** [comment]
@semriemens I opened an issue concerning the other bug you raised https://github.com/qgis/QGIS/issues/56022

**Pull request merged**
